CodeGen: Pass SubtargetInfo to TargetGenInstrInfo constructors

arsenm opened the pull request:
This will make it possible for tablegen to make subtarget
dependent decisions without adding new arguments to every
target.

arsenm committed:
CodeGen: Pass MCSubtargetInfo to TargetGenInstrInfo constructors

This will make it possible for tablegen to make subtarget
dependent decisions without adding new arguments to every
target.

s-barannikov commented:
Looks good as such, but can you provide more information? What kind of decisions and what new arguments?

Should the title mention TargetSubtargetInfo instead of MCSubtargetInfo?

Should it be propagated down to the base `TargetInstrInfo`? There are a lot of TII methods that need STI, but currently have to obtain it via MF / MI / MBB.

s-barannikov commented on llvm/lib/Target/BPF/BPFInstrInfo.h:
```diff
@@ -20,12 +20,13 @@
 
 public:
-  BPFInstrInfo();
+  BPFInstrInfo(const BPFSubtarget &STI);
```
```suggestion
  explicit BPFInstrInfo(const BPFSubtarget &STI);
```

s-barannikov commented on llvm/lib/Target/SPIRV/SPIRVInstrInfo.h:
```diff
@@ -20,12 +20,13 @@
 
 public:
-  SPIRVInstrInfo();
+  SPIRVInstrInfo(const SPIRVSubtarget &STI);
```
```suggestion
  explicit SPIRVInstrInfo(const SPIRVSubtarget &STI);
```

s-barannikov commented on llvm/lib/Target/XCore/XCoreInstrInfo.h:
```diff
@@ -20,12 +20,13 @@
   virtual void anchor();
 public:
-  XCoreInstrInfo();
+  XCoreInstrInfo(const XCoreSubtarget &ST);
```
```suggestion
  explicit XCoreInstrInfo(const XCoreSubtarget &ST);
```

s-barannikov commented on llvm/utils/TableGen/InstrInfoEmitter.cpp:
```diff
@@ -1070,7 +1070,9 @@ void InstrInfoEmitter::run(raw_ostream &OS) {
      << "  explicit " << ClassName
-     << "(unsigned CFSetupOpcode = ~0u, unsigned CFDestroyOpcode = ~0u, "
+     << "(const TargetSubtargetInfo &STI, unsigned CFSetupOpcode = ~0u, "
+        "unsigned "
+        "CFDestroyOpcode = ~0u, "
```
```suggestion
        "unsigned CFDestroyOpcode = ~0u, "
```

s-barannikov commented on llvm/utils/TableGen/InstrInfoEmitter.cpp:
```diff
@@ -1104,7 +1106,8 @@ void InstrInfoEmitter::run(raw_ostream &OS) {
   OS << ClassName << "::" << ClassName
-     << "(unsigned CFSetupOpcode, unsigned CFDestroyOpcode, unsigned "
+     << "(const TargetSubtargetInfo &STI, unsigned CFSetupOpcode, unsigned "
+        "CFDestroyOpcode, unsigned "
         "CatchRetOpcode, unsigned ReturnOpcode)\n"
```
```suggestion
        "CFDestroyOpcode, unsigned CatchRetOpcode, unsigned ReturnOpcode)\n"
```

s-barannikov reviewed (approved):
few nits

arsenm committed:
Apply suggestions from code review

Co-authored-by: Sergei Barannikov <barannikov88@gmail.com>

arsenm committed:
Update llvm/lib/Target/SPIRV/SPIRVInstrInfo.h

Co-authored-by: Sergei Barannikov <barannikov88@gmail.com>

arsenm committed:
Update llvm/lib/Target/BPF/BPFInstrInfo.h

Co-authored-by: Sergei Barannikov <barannikov88@gmail.com>

s-barannikov commented:
FWIW SPIRV issue seems pre-existing, I can reproduce it on main.

arsenm committed:
Work around SPIRV validation failure.

It may be relying on initializing InstrInfo with the under-initialized
subtarget.

arsenm committed:
Revert "Work around SPIRV validation failure."

This reverts commit 38b1d52ce49621c37f9dc43a1d78a0c1831de713.

arsenm commented:
HLSL fails with: Request POST /repos/llvm/llvm-project/check-runs failed with 403: Forbidden

SPIRV fails validation in a test that does not change output with this patch
